### PR TITLE
James 3755 OIDC check token by user info

### PR DESCRIPTION
**Pull request by vttranlina**
resolve https://github.com/linagora/james-project/issues/4686

### CODE REVIEW

File: server/protocols/jwt/src/main/java/org/apache/james/jwt/OidcJwtTokenVerifier.java [hunk 1]
```diff
@@ -65,17 +63,19 @@ public static <T> Optional<T> getClaimWithoutSignatureVerification(String token,
...
+            .flatMap(claimResult -> Mono.from(CHECK_TOKEN_CLIENT.introspect(introspectionEndpoint, jwtToken))
+                .filter(TokenIntrospectionResponse::active)
+                .map(activeResponse -> claimResult));
+    }
+
+    public static Publisher<String> verifyWithUserinfo(String jwtToken, URL jwksURL, String claimName, URL userinfoEndpoint) {
+        return Mono.fromCallable(() -> verifySignatureAndExtractClaim(jwtToken, jwksURL, claimName))
+            .flatMap(optional -> optional.map(Mono::just).orElseGet(Mono::empty))
+            .flatMap(claimResult -> Mono.from(CHECK_TOKEN_CLIENT.useInfo(userinfoEndpoint, jwtToken))
+                // .filter(userinfoResponse -> userinfoResponse.getPreferredUsername().equals(claimResult))
```
Comment (vttranlina) [inline server/protocols/jwt/src/main/java/org/apache/james/jwt/OidcJwtTokenVerifier.java:78]: Not sure here
If we want to check the `preferred_username` must be equal to `claimResult`, we need to add 1 more configure. 
Eg: `auth.oidc.userinfo.claimname`

**Comment by vttranlina** [comment]
[ ] add tests in `IMAPServerTest` and `SMTPSaslTest`

File: server/protocols/jwt/src/main/java/org/apache/james/jwt/OidcJwtTokenVerifier.java (hunk 1 above)
Comment (chibenwa) [reply server/protocols/jwt/src/main/java/org/apache/james/jwt/OidcJwtTokenVerifier.java:78]: Why not use `imapservers.imapserver.auth.oidc.claim` ?

Comment (vttranlina) [reply server/protocols/jwt/src/main/java/org/apache/james/jwt/OidcJwtTokenVerifier.java:78]: The first I see the 2 payloads are different, So a bit worried. 
But I rethinking again, It's ok. We can use `imapservers.imapserver.auth.oidc.claim`

File: server/protocols/jwt/src/main/java/org/apache/james/jwt/CheckTokenClient.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import org.apache.james.jwt.introspection.IntrospectionEndpoint;
+import org.apache.james.jwt.introspection.TokenIntrospectionResponse;
+import org.apache.james.jwt.userinfo.UserinfoResponse;
+import org.reactivestreams.Publisher;
+
+public interface CheckTokenClient {
+
+    Publisher<TokenIntrospectionResponse> introspect(IntrospectionEndpoint introspectionEndpoint, String token);
+
+    Publisher<UserinfoResponse> useInfo(URL userinfoEndpoint, String bearerToken);
```
Comment (Arsnael) [inline server/protocols/jwt/src/main/java/org/apache/james/jwt/CheckTokenClient.java:33]:
```suggestion
    Publisher<UserinfoResponse> userInfo(URL userinfoEndpoint, String bearerToken);
```

**Comment by Arsnael** [comment]
Also the fact that if introspection endpoint has been set up then it will be chosen over user info even if it's set up too, has it been discussed and agreed as well?

**Comment by vttranlina** [comment]
> Also the fact that if introspection endpoint has been set up then it will be chosen over user info even if it's set up too

Yes

**Comment by chibenwa** [comment]
IMO if both are configured, both should be called?

**Comment by vttranlina** [comment]
> IMO if both are configured, both should be called?

IMO, not necessary. 
// I checked on keycloak, we can't got userInfo with that token has been logout. 
The double call is un-necessary

**Comment by Arsnael** [comment]
Don't get me wrong it was a question :)

I agree with the fact that the introspection endpoint should always be preferred to the user info url if both are set up. I do believe actually that user info endpoint is nice in some cases like we saw, but the best practice is introspection endpoint. And calling both is redundant.

So I'm ok with that choice actually !

File: protocols/api/src/main/java/org/apache/james/protocols/api/OidcSASLConfiguration.java [hunk 3]
```diff
@@ -43,9 +43,16 @@ public static OidcSASLConfiguration parse(HierarchicalConfiguration<ImmutableNod
         Preconditions.checkNotNull(scope, "`scope` property need to be specified inside the oidc tag");
 
         String introspectionUrl = configuration.getString("introspection.url", null);
+        String userInfoUrl = configuration.getString("userinfo.url", null);
+
+        Preconditions.checkArgument(userInfoUrl == null
+                || userInfoUrl.startsWith("http://127.0.0.1")
```
[the remark this answers is not in the document]
Comment (vttranlina) [reply protocols/api/src/main/java/org/apache/james/protocols/api/OidcSASLConfiguration.java:49]: `"http://127.0.0.1"` just for test

Normally, the userInfo endpoint should support TLS 
```
Communication with the UserInfo Endpoint MUST utilize TLS. See Section 16.17 for more information on using TLS. 
(https://openid.net/specs/openid-connect-core-1_0.html#UserInfo)
```

but if it is a "internal request", should we accept HTTP?

[the remark this answers is not in the document]
Comment (quantranhong1999) [reply protocols/api/src/main/java/org/apache/james/protocols/api/OidcSASLConfiguration.java:49]: Well, I guess it is not even if it is an "internal request".

[the remark this answers is not in the document]
Comment (quantranhong1999) [reply protocols/api/src/main/java/org/apache/james/protocols/api/OidcSASLConfiguration.java:49]: > "http://127.0.0.1" just for test

Any possible vulnerability e.g a scenario on same localhost of James there is also a running Keycloak, then attackers could use this `userinfo.url` to probe/attack that Keycloak?

[the remark this answers is not in the document]
Comment (quantranhong1999) [reply protocols/api/src/main/java/org/apache/james/protocols/api/OidcSASLConfiguration.java:49]: Well forget about the vulnerability comment, this `userInfo.url` can not be settled by clients so not a concern.

[the remark this answers is not in the document]
Comment (chibenwa) [reply protocols/api/src/main/java/org/apache/james/protocols/api/OidcSASLConfiguration.java:49]: IMO we can consider the james configuration as "safe" and assume attackers can't modify it...

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
Can you squash, please?

**Pull request merged**